scoring/score.py

```diff
@@ -13,21 +24,182 @@ def __init__(self, message: str, *, code: str) -> None:
         self.code = code
 
 
+def join_text(strings: Iterable[str], separator: str) -> str:
+    """
+    Construct an english-language comma separated list ending with the given
+    separator word. This enables constructs like "foo, bar and baz" given a list
+    of names.
+
+    >>> join_text(["foo", "bar", "baz"], "and")
+    "foo, bar and baz"
+
+    >>> join_text(["foo", "bar"], "or")
+    "foo or bar"
+
+    >>> join_text(["foo"], "or")
+    "foo"
+    """
+    strings = tuple(strings)
+
+    try:
+        *strings, right = strings
+    except ValueError:
+        return ""
+
+    if not strings:
+        return right
+
+    left = ", ".join(strings[:-1])
+    return f"{left} {separator} {strings[-1]}"
+
+
+def join_and(strings: Iterable[str]) -> str:
+    return join_text(strings, "and")
+
+
+def join_or(strings: Iterable[str]) -> str:
+    return join_text(strings, "or")
+
+
 class Scorer:
     def __init__(self, teams_data, arena_data):
         self._teams_data = teams_data
-        self._arena_data = arena_data
+        self._districts = arena_data['other']['districts']
+
+        for district in self._districts.values():
+            district['pallet_counts'] = collections.Counter(
+                district['pallets'].replace(' ', ''),
+            )
+            district['highest'] = district['highest'].replace(' ', '')
+
+    def score_district_for_zone(self, name: str, district: District, zone: int) -> int:
+        colour = ZONE_COLOURS[zone]
+
+        num_tokens = district['pallet_counts'][colour]
+        score = num_tokens * DISTRICT_SCORE_MAP[name]
+
+        if colour in district['highest']:
+            # Points are doubled for the team owning the highest pallet
+            score *= 2
+
+        return score
 
     def calculate_scores(self):
         scores = {}
 
         for tla, info in self._teams_data.items():
-            scores[tla] = 0
+            district_score = sum(
+                self.score_district_for_zone(name, district, info['zone'])
+                for name, district in self._districts.items()
+            )
+            movement_score = 1 if info.get('left_starting_zone') else 0
+            scores[tla] = district_score + movement_score
 
         return scores
 
     def validate(self, other_data):
-        warnings.warn("Scoresheet validation not implemented")
+        # Check that the right districts are specified.
+        if self._districts.keys() != DISTRICT_SCORE_MAP.keys():
+            missing = DISTRICT_SCORE_MAP.keys() - self._districts.keys()
+            extra = self._districts.keys() - DISTRICT_SCORE_MAP.keys()
+            detail = "Wrong districts specified."
+            if missing:
+                detail += f" Missing: {join_and(missing)}."
+            if extra:
+                detail += f" Extra: {join_and(repr(x) for x in extra)}."
+            raise InvalidScoresheetException(
+                detail,
+                code='invalid_districts',
+            )
```

Comment:
What does this test do? Where are the districts specified?

Reply:
This checks that the districts actually present are the expected ones. The districts are specified by `DISTRICT_SCORE_MAP` which is defined within the software (specifically it's imported from `sr2025.py`).